src/lib.rs

```diff
@@ -334,6 +335,212 @@ where
     }
 }
 
+#[cfg(feature = "async")]
+impl<I2C> Icm42670<I2C>
+where
+    I2C: embedded_hal_async::i2c::I2c,
+{
+    /// Instantiate a new instance of the driver and initialize the device
+    pub async fn new(i2c: I2C, address: Address) -> Result<Self, Error<I2C::Error>> {
+        let mut me = Self { i2c, address };
+
+        // Verify that the device has the correct ID before continuing. If the ID does
+        // not match either of the expected values then it is likely the wrong chip is
+        // connected.
+        if !Self::DEVICE_IDS.contains(&me.device_id_async().await?) {
+            return Err(Error::SensorError(SensorError::BadChip));
+        }
+
+        // Make sure that any configuration has been restored to the default values when
+        // initializing the driver.
+        me.set_accel_range_async(AccelRange::default()).await?;
+        me.set_gyro_range_async(GyroRange::default()).await?;
+
+        // The IMU uses `PowerMode::Sleep` by default, which disables both the accel and
+        // gyro, so we enable them both during driver initialization.
+        me.set_power_mode_async(PowerMode::SixAxisLowNoise).await?;
+
+        Ok(me)
+    }
+
+    /// Read the ID of the connected device
+    pub async fn device_id_async(&mut self) -> Result<u8, Error<I2C::Error>> {
+        self.read_reg_async(&Bank0::WHO_AM_I).await
+    }
+
+    /// Perform a software-reset on the device
+    pub async fn soft_reset_async(&mut self) -> Result<(), Error<I2C::Error>> {
+        self.update_reg_async(&Bank0::SIGNAL_PATH_RESET, 0x10, 0b0001_0000).await
+    }
```

Comment:
If you try running the non-async version in release mode, then imediately go into a read-loop, you'll get bogus data. The datasheet isn't clear what the reset-time needs to be, but 100ms seems to work for me. A more refined api would leave it up to the user to a) opt out of the delay entirely and b) choose their own delay.

(written for illustrative purposes only)

```suggestion
    pub async fn soft_reset_async(&mut self, delayer: &mut impl embedded_hal_async::delay::DelayUs) -> Result<(), Error<I2C::Error>> {
        let res = self.update_reg_async(&Bank0::SIGNAL_PATH_RESET, 0x10, 0b0001_0000).await;
        // The IMU needs some time to reset.
        delayer.delay_ms(100).await;
        res
    }
```

In the non-async version, it could make sense to just document the requirement, but with async, there's a bit of an "agreement" that something could block, no?

Reply:
I would leave the impl simpler by just specifying a set delay that works out of the box.
For some driver crates I've seen something like 10ms working well.

I will be able to test this once I can reliably work with the sensor...